### PR TITLE
#39 importing categories

### PR DESCRIPTION
**Pull request by nickmoreton**
Codebase: https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/39

Rebasing didn't work so I had to cherry pick the commits in. I expect that's because of the earlier refactor work.

### CODE REVIEW

File: docs/categories.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+It is disabled by default but you can enable the plugin by adding the following to your settings file.
+
+```python
+WAGTAIL_WORDPRESS_IMPORT_CATEGORY_PLUGIN_ENABLED = True
+```
+
+The plugin will also need to know which model in you app is used for the categories so add the following to your settings file.
+
+```python
+WAGTAIL_WORDPRESS_IMPORT_CATEGORY_PLUGIN_MODEL = "pages.models.Category
```
Comment (nimasmi) [inline docs/categories.md]: **Suggestion**: fix typo

```suggestion
WAGTAIL_WORDPRESS_IMPORT_CATEGORY_PLUGIN_MODEL = "pages.models.Category"
```

File: docs/categories.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+```
+
+**Name field minimum length**: We implemented this because in testing we found that many categories of a single character in length. Name's that are only a single character aren't imported.
```
Comment (nimasmi) [inline docs/categories.md]: **Suggestion**: reword

```suggestion
**Name field minimum length**: Some WordPress sites can have categories just a single character in length. If these are not useful, they can be skipped using this setting.
```

File: docs/categories.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+You can alter the minimum length to suit your needs by adding the following to your settings file.
+
+```python
+WAGTAIL_WORDPRESS_IMPORT_CATEGORY_PLUGIN_MIN_NAME_LENGTH = [num] as an integer
```
Comment (nimasmi) [inline docs/categories.md]: **Suggestion**: Use syntactically correct example code

```suggestion
WAGTAIL_WORDPRESS_IMPORT_CATEGORY_PLUGIN_MIN_NAME_LENGTH = 3  # for example
```

File: example/models.py [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
 from wagtail_wordpress_import.blocks import WPImportStreamBlocks
+from wagtail_wordpress_import.models import WPImportedPageMixin
 
 
 class TestPage(WPImportedPageMixin, Page):
     body = StreamField(WPImportStreamBlocks)
-
+    categories = ParentalManyToManyField("example.Category", blank=True)
     content_panels = Page.content_panels + [
+        # FieldPanel("categories", widget=forms.CheckboxSelectMultiple),
```
Comment (nimasmi) [inline example/models.py]: **Question**: Why is this commented out? It looks like it's a necessary part of the model definition.

File: wagtail_wordpress_import/importers/wordpress.py [hunk 5]
```diff
@@ -207,6 +218,20 @@ def get_page(self, link, page):
         except self.page_model_instance.DoesNotExist:
             pass
 
+    def connect_page_categories(self, page, category_model, item):
+        categories = item.get("category")
+        if categories:
+            page_categories = []
```
Comment (nimasmi) [inline wagtail_wordpress_import/importers/wordpress.py:229]: **Question**: We reset categories to an empty list first here — would we ever be reimporting the data, or importing categories for an existing page?

File: wagtail_wordpress_import/importers/wordpress.py [hunk 6]
```diff
@@ -207,6 +218,20 @@ def get_page(self, link, page):
...
+            for name in categories:
+                if len(name) > category_name_min_length():
+                    existing_category = None
+                    try:
+                        existing_category = category_model.objects.get(name=name)
+                    except category_model.DoesNotExist:
+                        existing_category = category_model.objects.create(name=name)
+                    page_categories.append(existing_category)
```
Comment (nimasmi) [inline wagtail_wordpress_import/importers/wordpress.py]: **Suggestion**: Use `.get_or_create()`

```suggestion
                    category, created = category_model.objects.get_or_create(name=name)
                    page_categories.append(category)
```

File: wagtail_wordpress_import/test/fixtures/raw_xml.xml [hunk 7]
```diff
@@ -170,6 +172,8 @@ Nihil hic munitissimus habendi senatus locus, nihil horum?.&lt;/blockquote&gt;</
             <wp:post_type>page</wp:post_type>
             <wp:post_password/>
             <wp:is_sticky>0</wp:is_sticky>
+            <category domain="category" nicename="blogging">Cars</category>
+            <category domain="category" nicename="life">Computing</category>
```
Comment (nimasmi) [inline wagtail_wordpress_import/test/fixtures/raw_xml.xml]: **Comment**: The `nicename` attributes don't correspond to the displayed text.

File: wagtail_wordpress_import/test/tests/test_wordpress_item.py [hunk 8]
```diff
@@ -82,3 +95,45 @@ def test_cleaned_fields(self):
...
+        self.assertEqual(len(snippets), 4)
+
+    def test_page_has_categories(self):
+        page_one = self.imported_pages.get(title="Item one title")
+        page_one_categories = page_one.specific.categories.all()
+        self.assertEqual(2, page_one_categories.count())
+
+        page_two = self.imported_pages.get(title="Item two title")
+        page_two_categories = page_two.specific.categories.all()
+        self.assertEqual(2, page_two_categories.count())
```
Comment (nimasmi) [inline wagtail_wordpress_import/test/tests/test_wordpress_item.py]: **Suggestion**: This looks like a separate assertion. Can we test it in a separate test case?

**Suggestion**: Add a test of what the categories are, e.g. "cars", "life" etc.

File: wagtail_wordpress_import/test/tests/test_wordpress_item.py [hunk 9]
```diff
@@ -82,3 +95,45 @@ def test_cleaned_fields(self):
...
+        self.imported_pages = self.parent_page.get_children().all()
+
+    def test_category_snippets_are_saved(self):
+        snippets = Category.objects.all()
+        self.assertEqual(len(snippets), 4)
+
+    def test_page_has_categories(self):
+        page_one = self.imported_pages.get(title="Item one title")
+        page_one_categories = page_one.specific.categories.all()
+        self.assertEqual(2, page_one_categories.count())
```
Comment (nimasmi) [inline wagtail_wordpress_import/test/tests/test_wordpress_item.py]: **Suggestion**: Add a test of what the categories are, e.g. "cars", "life" etc.

File: wagtail_wordpress_import/test/tests/test_wordpress_item.py (hunk 8 above)
Comment (nimasmi) [inline wagtail_wordpress_import/test/tests/test_wordpress_item.py]: **Suggestion**: Add a test that short category names are ignored if below the min length setting.

Comment (nimasmi) [inline wagtail_wordpress_import/test/tests/test_wordpress_item.py]: **Suggestion**: Add a test that multiple pages sharing a category don't result in duplicate category instances being created.

File: example/models.py [hunk 10]
```diff
@@ -34,3 +37,17 @@ def import_wordpress_data(self, data):
 
         # own model fields
         self.body = data["body"]
+
+
+@register_snippet
+class Category(models.Model):
+    name = models.CharField(max_length=255)
```
Comment (nimasmi) [inline example/models.py]: **Suggestion**: We're using this as a "get_or_create" field. It would be good to add a uniqueness constraint.

File: example/models.py (hunk 4 above)
Comment (nickmoreton) [reply example/models.py]: It's not required in the mixin i should have removed it. 

In the budgets are sexy provided wagtail app they use the promote panel for the categories and use the CheckboxSelectMultiple widget. The developer would make that decision.

File: wagtail_wordpress_import/test/fixtures/raw_xml.xml (hunk 7 above)
Comment (nickmoreton) [reply wagtail_wordpress_import/test/fixtures/raw_xml.xml]: Thanks I will correct it.

File: wagtail_wordpress_import/importers/wordpress.py (hunk 5 above)
Comment (nickmoreton) [reply wagtail_wordpress_import/importers/wordpress.py:229]: I made some changes to this function possibly a little more than you mention as I found an edge case where if no categories are attached to a page it could fail.

**Comment by nimasmi** [comment]
Thanks. I approve of these changes.
